### ISSUE TITLE
ARM64/Graviton support

### ISSUE THREAD

**Issue by jasonmccallister** [brefphp/bref#1043]
AWS launched ARM64 support on Lambda (https://aws.amazon.com/blogs/aws/aws-lambda-functions-powered-by-aws-graviton2-processor-run-your-functions-on-arm-and-get-up-to-34-better-price-performance/).

Would be useful to get Bref functions and container images setup with ARM64 variants... Happy to assist where I can.

**Comment by mnapoli** [comment]
AFAIK PHP isn't running best on ARM (e.g. JIT not supported). I may be wrong, but it might be worth digging into that?

**Comment by deleugpn** [comment]
I think the first step is to wait Cloudformation support? https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/926

**Comment by jasonmccallister** [comment]
@deleugpn for Lambda layers yes, but I think there should not be anything holding up pushing the ARM images to Docker Hub for container lambda functions?

**Comment by dytyniuk** [comment]
@deleugpn people claim in https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/926#issuecomment-931785524 that CloudFormation API already supports the `Architecture` property.

@mnapoli maybe, this helps to melt your concern about running PHP on ARM -- https://aws.amazon.com/tr/blogs/compute/improving-performance-of-php-for-arm64-and-impact-on-amazon-ec2-m6g-instances/

**Comment by mnapoli** [comment]
@dytyniuk definitely, thank you. I was wrong, ARM can be interesting for PHP.

Also relevant: https://twitter.com/sebastienhouze/status/1443742924534132843

> And that ARM JIT support is coming at least with 8.1

Also FYI: CloudFormation support is in (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-architectures) and we're planning to add support in Serverless Framework early next week.

**Comment by willvincent** [comment]
So if the roadblocks to this were cloudformation and the serverless framework supporting arm - both of which now do.. is there any remaining issue or is this good to go now? And if so.. how so? 😁

I'm about to dig into a new project and I think I'd like to use the graviton lambdas if possible.

**Comment by dytyniuk** [comment]
Hey @deleugpn 
Maybe, this repository can help -- https://repo.drpixel.fr/enterprise/7/drpixel-php74/aarch64/ ?

**Comment by deleugpn** [comment]
I managed to install php with `amazon-linux-extras` but its locked into php 8.0.8 which is a few months behind. I'm not sure it would be a good switch as it would delay patch releases beyond our control. If anyone have a Dockerfile recipe on how to install php using the `al2-arm64` base image and wants to share, I'd be grateful!

**Comment by deleugpn** [comment]
Another thing that might be relevant, if I understand https://repo.drpixel.fr/ correctly, it only support PHP 8.0.1, which is 10 point-releases behind so even if I did manage to install with it, it seems it would be very slow to move forward with PHP upgrades. I'm open to suggestions and considering our options. I might have to dig deep into our current compile php ourselves and abort the attempt to simplify it.

**Comment by dytyniuk** [comment]
@deleugpn yeah, now I feel your shoes. At least, to some extent.

IMHO, any effort to reuse pre-built binaries for this or that repository will bring a concrete PHP version and lock further updates. From this point of view, layers' updates will be postponed until maintainers of the repository incorporate an update.

Despite the pain of the compilation, I think, this is the only way to provide best layers.

**Comment by dytyniuk** [comment]
@deleugpn okay, I understand there's a way to extract an artefact from _any_ place in the container. But, is an explicit listing more convenient rather than copying every binary/module installed on a build time? Imagine you have an auxiliary module of a file built for an extension, and it's missing from your list. How long will it take to pinpoint a root cause of a broken extension?

**Comment by georgeboot** [comment]
How can I help getting this shipped?

**Comment by deleugpn** [comment]
I've been working on https://github.com/brefphp/bref/pull/1078 for quite a while now but docker-compose bug on ARM prevents it from building arm layers at the moment. In the meantime I'm trying to finish up how to integrate the console layer into the new proposal so that we can start beta test it while we wait on https://github.com/docker/compose/issues/8804

**Comment by allan-simon** [comment]
@deleugpn  interested in testing the x86 layers

**Comment by deleugpn** [comment]
arn:aws:lambda:us-east-1:978790411843:layer:prototype-arm64-php74-function:1
 arn:aws:lambda:us-east-1:179453031647:layer:prototype-arm64-php74-fpm:1

**Comment by allan-simon** [comment]
@deleugpn  do you have them in 8.0 by chance ? our code base use some 8.x-only features :)

**Comment by deleugpn** [comment]
I will look into building it this week. Sorry for the short answer, I'm on holidays between Christmas and new years

**Comment by allan-simon** [comment]
no problem, enjoy your vacations :)

**Comment by deleugpn** [comment]
sorry the long time without replies. I have actually tried to release ARM64 prototype layers across regions multiple times in the past 3 weeks and no luck as of yet.

Some exotic AWS regions seems to be bugged when trying to publish Graviton layers because it always gives "The security token in the request is invalid" even though the build process is a perfect replica between x86 and arm64. I tried disabling these regions and now there is an unknown error crashing the build process just after the layers are published which I'm still unable to track down. In the meantime, docker-compose is still bugged on arm64 when trying to build using buildkit, so builds are much slower for arm64. 

Unfortunately the progress here is a bit slow as there's a lot of moving parts that are still not working well with arm64 and sometimes don't even give clear errors.

**Comment by allan-simon** [comment]
thanks for keeping us updated , especially as it looks like a very tedious and "death by a thousand cut" work , hope you will succeed :)

**Comment by jasonmccallister** [comment]
@deleugpn I'm going to have some focus time at work to dedicate to this, happy to collab on what you have done so far and help push this over the finish line.

**Comment by wlami** [comment]
> Some exotic AWS regions seems to be bugged when trying to publish Graviton layers because it always gives "The security token in the request is invalid" even though the build process is a perfect replica between x86 and arm64.

Please note that Graviton2 processor support is not available in all regions ([see](https://aws.amazon.com/blogs/aws/aws-lambda-functions-powered-by-aws-graviton2-processor-run-your-functions-on-arm-and-get-up-to-34-better-price-performance/)): 
> You can use Lambda Functions powered by Graviton2 processor today in US East (N. Virginia), US East (Ohio), US West (Oregon), Europe (Frankfurt), Europe (Ireland), EU (London), Asia Pacific (Mumbai), Asia Pacific (Singapore), Asia Pacific (Sydney), Asia Pacific (Tokyo).

@deleugpn  Are there any errors when deploying in the supported regions?

**Comment by deleugpn** [comment]
@wlami Thank you very much for that list! Unfortunately I just tried that and it did not solve the issue https://github.com/deleugpn/bref/pull/53

**Comment by deleugpn** [comment]
If anybody can figure out why this script fails at the end, please I'm open to any ideas 

<details>
    <summary>AWS CodeBuild Logs</summary>

```
Successfully built 1f0adeee2251
Successfully tagged bref/arm-php80-fpm-zip:latest
Network runtime_default  Creating
Network runtime_default  Created
Container runtime-php80-zip-fpm-1  Creating
Container runtime-php74-zip-fpm-1  Creating
Container runtime-php74-zip-function-1  Creating
Container runtime-php80-zip-function-1  Creating
Container runtime-php80-zip-fpm-1  Created
Container runtime-php80-zip-function-1  Created
Container runtime-php74-zip-fpm-1  Created
Container runtime-php74-zip-function-1  Created
Attaching to runtime-php74-zip-fpm-1, runtime-php74-zip-function-1, runtime-php80-zip-fpm-1, runtime-php80-zip-function-1
runtime-php74-zip-fpm-1 exited with code 0
runtime-php74-zip-function-1 exited with code 0
runtime-php80-zip-fpm-1 exited with code 0
runtime-php80-zip-function-1 exited with code 0
Container runtime-php80-zip-fpm-1  Stopping
Container runtime-php80-zip-fpm-1  Stopping
Container runtime-php74-zip-fpm-1  Stopping
Container runtime-php80-zip-function-1  Stopping
Container runtime-php74-zip-function-1  Stopping
Container runtime-php74-zip-function-1  Stopping
Container runtime-php80-zip-function-1  Stopping
Container runtime-php74-zip-fpm-1  Stopping
Container runtime-php80-zip-fpm-1  Stopped
Container runtime-php80-zip-fpm-1  Removing
Container runtime-php74-zip-fpm-1  Stopped
Container runtime-php74-zip-fpm-1  Removing
Container runtime-php74-zip-function-1  Stopped
Container runtime-php74-zip-function-1  Removing
Container runtime-php80-zip-function-1  Stopped
Container runtime-php80-zip-function-1  Removing
Container runtime-php80-zip-fpm-1  Removed
Container runtime-php80-zip-function-1  Removed
Container runtime-php74-zip-fpm-1  Removed
Container runtime-php74-zip-function-1  Removed
Network runtime_default  Removing
Network runtime_default  Removed
make[1]: Entering directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
# Add executable permission to the publish script
chmod +x ./publish.sh
LAYER_NAME=arm-php74-function make parallel-publish
make[2]: Entering directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
REGION=us-east-1 ./publish.sh #US East (N. Virginia)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 12 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-east-1:179453031647:layer:prototype-arm-php74-function:12"}
[Publish] Layer  added!
REGION=us-east-2 ./publish.sh #US East (Ohio)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 12 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-east-2:179453031647:layer:prototype-arm-php74-function:12"}
[Publish] Layer  added!
#REGION=us-west-1 ./publish.sh #US West (N. California)
REGION=us-west-2 ./publish.sh #US West (Oregon)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 12 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-west-2:179453031647:layer:prototype-arm-php74-function:12"}
[Publish] Layer  added!
#REGION=ca-central-1 ./publish.sh #Canada (Central)
#REGION=sa-east-1 ./publish.sh #South America (São Paulo)
REGION=eu-west-1 ./publish.sh #Europe (Ireland)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 12 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-west-1:179453031647:layer:prototype-arm-php74-function:12"}
[Publish] Layer  added!
REGION=eu-west-2 ./publish.sh #Europe (London)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 12 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-west-2:179453031647:layer:prototype-arm-php74-function:12"}
[Publish] Layer  added!
#REGION=eu-west-3 ./publish.sh #Europe (Paris)
#REGION=eu-north-1 ./publish.sh #Europe (Stockholm)
#REGION=eu-south-1 ./publish.sh #Europe (Milan)
REGION=eu-central-1 ./publish.sh #Europe (Frankfurt)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-central-1:179453031647:layer:prototype-arm-php74-function:4"}
[Publish] Layer  added!
#REGION=ap-east-1 ./publish.sh #Asia Pacific (Hong Kong)
REGION=ap-south-1 ./publish.sh #Asia Pacific (Mumbai)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-south-1:179453031647:layer:prototype-arm-php74-function:4"}
[Publish] Layer  added!
REGION=ap-southeast-1 ./publish.sh #Asia Pacific (Singapore)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-southeast-1:179453031647:layer:prototype-arm-php74-function:4"}
[Publish] Layer  added!
REGION=ap-northeast-1 ./publish.sh #Asia Pacific (Tokyo)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 12 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-northeast-1:179453031647:layer:prototype-arm-php74-function:12"}
[Publish] Layer  added!
#REGION=ap-northeast-3 ./publish.sh #Asia Pacific (Osaka)
#REGION=ap-northeast-2 ./publish.sh #Asia Pacific (Seoul)
#REGION=af-south-1 ./publish.sh #Africa (Cape Town)
#REGION=me-south-1 ./publish.sh #Middle East (Bahrain)
REGION=ap-southeast-2 ./publish.sh #Asia Pacific (Sydney)
[Publish] Publishing layer arm-php74-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-southeast-2:179453031647:layer:prototype-arm-php74-function:4"}
[Publish] Layer  added!
make[2]: Leaving directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
make[1]: Leaving directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
make[1]: Entering directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
# Add executable permission to the publish script
chmod +x ./publish.sh
LAYER_NAME=arm-php80-function make parallel-publish
make[2]: Entering directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
REGION=us-east-1 ./publish.sh #US East (N. Virginia)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-east-1:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
REGION=us-east-2 ./publish.sh #US East (Ohio)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-east-2:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
#REGION=us-west-1 ./publish.sh #US West (N. California)
REGION=us-west-2 ./publish.sh #US West (Oregon)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-west-2:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
#REGION=ca-central-1 ./publish.sh #Canada (Central)
#REGION=sa-east-1 ./publish.sh #South America (São Paulo)
REGION=eu-west-1 ./publish.sh #Europe (Ireland)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-west-1:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
REGION=eu-west-2 ./publish.sh #Europe (London)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-west-2:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
#REGION=eu-west-3 ./publish.sh #Europe (Paris)
#REGION=eu-north-1 ./publish.sh #Europe (Stockholm)
#REGION=eu-south-1 ./publish.sh #Europe (Milan)
REGION=eu-central-1 ./publish.sh #Europe (Frankfurt)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-central-1:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
#REGION=ap-east-1 ./publish.sh #Asia Pacific (Hong Kong)
REGION=ap-south-1 ./publish.sh #Asia Pacific (Mumbai)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-south-1:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
REGION=ap-southeast-1 ./publish.sh #Asia Pacific (Singapore)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-southeast-1:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
REGION=ap-northeast-1 ./publish.sh #Asia Pacific (Tokyo)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-northeast-1:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
#REGION=ap-northeast-3 ./publish.sh #Asia Pacific (Osaka)
#REGION=ap-northeast-2 ./publish.sh #Asia Pacific (Seoul)
#REGION=af-south-1 ./publish.sh #Africa (Cape Town)
#REGION=me-south-1 ./publish.sh #Middle East (Bahrain)
REGION=ap-southeast-2 ./publish.sh #Asia Pacific (Sydney)
[Publish] Publishing layer arm-php80-function...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-southeast-2:179453031647:layer:prototype-arm-php80-function:4"}
[Publish] Layer  added!
make[2]: Leaving directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
make[1]: Leaving directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
make[1]: Entering directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
# Add executable permission to the publish script
chmod +x ./publish.sh
LAYER_NAME=arm-php74-fpm make parallel-publish
make[2]: Entering directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
REGION=us-east-1 ./publish.sh #US East (N. Virginia)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-east-1:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
REGION=us-east-2 ./publish.sh #US East (Ohio)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-east-2:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
#REGION=us-west-1 ./publish.sh #US West (N. California)
REGION=us-west-2 ./publish.sh #US West (Oregon)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-west-2:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
#REGION=ca-central-1 ./publish.sh #Canada (Central)
#REGION=sa-east-1 ./publish.sh #South America (São Paulo)
REGION=eu-west-1 ./publish.sh #Europe (Ireland)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-west-1:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
REGION=eu-west-2 ./publish.sh #Europe (London)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-west-2:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
#REGION=eu-west-3 ./publish.sh #Europe (Paris)
#REGION=eu-north-1 ./publish.sh #Europe (Stockholm)
#REGION=eu-south-1 ./publish.sh #Europe (Milan)
REGION=eu-central-1 ./publish.sh #Europe (Frankfurt)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-central-1:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
#REGION=ap-east-1 ./publish.sh #Asia Pacific (Hong Kong)
REGION=ap-south-1 ./publish.sh #Asia Pacific (Mumbai)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-south-1:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
REGION=ap-southeast-1 ./publish.sh #Asia Pacific (Singapore)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-southeast-1:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
REGION=ap-northeast-1 ./publish.sh #Asia Pacific (Tokyo)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-northeast-1:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
#REGION=ap-northeast-3 ./publish.sh #Asia Pacific (Osaka)
#REGION=ap-northeast-2 ./publish.sh #Asia Pacific (Seoul)
#REGION=af-south-1 ./publish.sh #Africa (Cape Town)
#REGION=me-south-1 ./publish.sh #Middle East (Bahrain)
REGION=ap-southeast-2 ./publish.sh #Asia Pacific (Sydney)
[Publish] Publishing layer arm-php74-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-southeast-2:179453031647:layer:prototype-arm-php74-fpm:4"}
[Publish] Layer  added!
make[2]: Leaving directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
make[1]: Leaving directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
make[1]: Entering directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
# Add executable permission to the publish script
chmod +x ./publish.sh
LAYER_NAME=arm-php80-fpm make parallel-publish
make[2]: Entering directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
REGION=us-east-1 ./publish.sh #US East (N. Virginia)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-east-1:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
REGION=us-east-2 ./publish.sh #US East (Ohio)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-east-2:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
#REGION=us-west-1 ./publish.sh #US West (N. California)
REGION=us-west-2 ./publish.sh #US West (Oregon)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:us-west-2:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
#REGION=ca-central-1 ./publish.sh #Canada (Central)
#REGION=sa-east-1 ./publish.sh #South America (São Paulo)
REGION=eu-west-1 ./publish.sh #Europe (Ireland)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-west-1:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
REGION=eu-west-2 ./publish.sh #Europe (London)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-west-2:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
#REGION=eu-west-3 ./publish.sh #Europe (Paris)
#REGION=eu-north-1 ./publish.sh #Europe (Stockholm)
#REGION=eu-south-1 ./publish.sh #Europe (Milan)
REGION=eu-central-1 ./publish.sh #Europe (Frankfurt)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:eu-central-1:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
#REGION=ap-east-1 ./publish.sh #Asia Pacific (Hong Kong)
REGION=ap-south-1 ./publish.sh #Asia Pacific (Mumbai)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-south-1:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
REGION=ap-southeast-1 ./publish.sh #Asia Pacific (Singapore)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-southeast-1:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
REGION=ap-northeast-1 ./publish.sh #Asia Pacific (Tokyo)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-northeast-1:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
#REGION=ap-northeast-3 ./publish.sh #Asia Pacific (Osaka)
#REGION=ap-northeast-2 ./publish.sh #Asia Pacific (Seoul)
#REGION=af-south-1 ./publish.sh #Africa (Cape Town)
#REGION=me-south-1 ./publish.sh #Middle East (Bahrain)
REGION=ap-southeast-2 ./publish.sh #Asia Pacific (Sydney)
[Publish] Publishing layer arm-php80-fpm...
[Publish] Layer 4 published! Adding layer permission...
{"Sid":"public","Effect":"Allow","Principal":"*","Action":"lambda:GetLayerVersion","Resource":"arn:aws:lambda:ap-southeast-2:179453031647:layer:prototype-arm-php80-fpm:4"}
[Publish] Layer  added!
make[2]: Leaving directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
make[1]: Leaving directory `/codebuild/output/src031647930/src/github.com/deleugpn/bref/runtime/common/publish'
Network utils_default  Creating
Network utils_default  Created
make: *** [everything] Error 255

[Container] 2022/01/31 22:10:14 Command did not exit successfully make -f cpu-${CPU}.Makefile everything exit status 2
[Container] 2022/01/31 22:10:14 Phase complete: BUILD State: FAILED_WITH_ABORT
[Container] 2022/01/31 22:10:14 Phase context status code: COMMAND_EXECUTION_ERROR Message: Error while executing command: make -f cpu-${CPU}.Makefile everything. Reason: exit status 2
```
</details>

Real Logs:
https://us-east-1.codebuild.aws.amazon.com/project/eyJlbmNyeXB0ZWREYXRhIjoib25lTkcvanI1K3FsbnBtN1NkeCtnYWRCWFFLQ1IrSHBYUllLVTRYV3I4S1B6RUU5UmJCRHUyb29NcjZ1bUV3YmVNMmUzRTd3RHd5OVhrRmVzdmFTMmZ3SFdxZHQwZHJ2bjJNbjBJRE43Vkd6NmVQeW00bVZDZ2NPVUhUQkhjSldZWDQ9IiwiaXZQYXJhbWV0ZXJTcGVjIjoiOHFqRUFlYkIyVUtXUVduUyIsIm1hdGVyaWFsU2V0U2VyaWFsIjoxfQ%3D%3D/build/0a3c357c-bb2f-45d2-ab8e-78d3b9368b18

**Comment by mattrenner** [comment]
Thanks very much for your work on this @deleugpn 

Given we see "Network utils_default  Created", I assume the build got as far in the makefile as 

	docker-compose -f common/utils/docker-compose.yml run parse

I suspect a return code of 255 suggests a PHP exception is being thrown by `parse-output-into-layers-json.php`. 

More specifically, should [line 18](https://github.com/deleugpn/bref/blob/e0e0c5d226ab623585c845b26259779b452e8721/runtime/common/utils/parse-output-into-layers-json.php#L18) be looking for $cpu to equal `arm` instead of `arm64`? 

	} elseif ($cpu === 'arm64') {

The environment variable set in CodeBuild has CPU=arm

**Comment by jasonmccallister** [comment]
I think some platforms output that  as aarch64?

**Comment by deleugpn** [comment]
@mattrenner I'm not even sure if I'm mad or hopeful (or both), but I'm running a new build fixing that. If that is the case, I still would like to try to understand why the error message doesn't show up (it would have saved me 1 month of trial-and-error).

In any case I will update shortly

**Comment by deleugpn** [comment]
thank you so much @mattrenner you have no idea how much that helped. Hopefully this will unblock the progress on having some ARM Layers for testing.

The pull request sent by the bot should have the ARN for the Layers to be tested out.

https://github.com/brefphp/bref/pull/1155

**Comment by deleugpn** [comment]
If anyone have any feedback on ARM layers, feel free to send them my way on our Slack, this issue or even at deleugyn at gmail.

**Comment by RobinHoutevelts** [comment]
We are running Bref in development on lambda using Docker. I'm willing to try this out, how can I change to ARM image?

Our current docker image:

```dockerfile
FROM bref/php-81-fpm

COPY --from=bref/extra-redis-php-81 /opt/bref-extra /opt/bref-extra

COPY . /var/task

CMD ["public/index.php"]
```

**Comment by CodeByTwo** [comment]
Any update on this? Would like to be able to run my serverless Laravel application on arm64 Lambdas utilising the graviton processors for speed and cost improvements! Thanks

**Comment by aran112000** [comment]
Morning @mnapoli, @deleugpn. Is ARM/Graviton support something likely to be finalised and added to Bref, feels like interest may have tailed off on the PRs around this?

**Comment by mnapoli** [comment]
Hey hey! Actually this is great timing, I'm working on that at the moment! I have picked up #1078. I cannot provide any estimate, but I'd like to have a beta out by the end of the month.

**Comment by mnapoli** [comment]
More details here: #1299

**Comment by mnapoli** [comment]
We have beta ARM layers ready to test 🎉 

I've started benchmarking them at https://github.com/brefphp/benchmarks/pull/5, but it benchmarks cold starts (which get slightly faster) and warm executions which are really fast anyway. So I'd love to get numbers from real applications, as well as see if these layers miss anything (e.g. extensions).

Anyone interested in testing ARM layers?

Here's how to proceed:

1. Upgrade to Bref v2 (dev version): `"bref/bref": "v2.x-dev"` in `composer.json` (and run `composer update`)
2. In `serverless.yml`, add `architecture: arm64` on the function you want to test (or add it in `provider:` to apply to all functions)
3. In `serverless.yml`, add `arm-` as a prefix for the layer, for example `${bref:layer.arm-php-80}` or `${bref:layer.arm-php-80-fpm}` (**note:** ARM layers only exist for PHP 8.0)
4. Deploy

I hope I haven't forgotten anything. Let me know if you see improvements to the execution time. If you can, please let us know how many invokes you tested with (if you tested with 10-100 invocations then the numbers might not be very reliable).

If you want to measure cold starts specifically, here is a CloudWatch Insights query you can run:

```
filter @type = “REPORT” and @initDuration
| stats
 count(@type) as count,
 min(@billedDuration) as min,
 avg(@billedDuration) as avg,
 pct(@billedDuration, 50) as p50,
 max(@billedDuration) as max
by @log
| sort @log
```

**Comment by mnapoli** [comment]
Here's a reminder that you can test ARM runtimes right now. Switch to v2 in `composer.json`:

```json
        "bref/bref": "v2.x-dev"
```

Then `composer update`. Update `serverless.yml` to use ARM runtimes:

```diff
functions:
    api:
        handler: public/index.php
+        architecture: arm64
        layers:
-            - ${bref:layer.php-80-fpm}
+            - ${bref:layer.arm-php-80-fpm}
        events:
            -   httpApi: '*'
```

There shouldn't be any big breaking change in v2 worth mentioning here ([upgrade guide here](https://github.com/brefphp/bref/blob/v2/docs/upgrading/v2.md)).

I am updating the benchmarks to include ARM results: https://github.com/brefphp/benchmarks/pull/5 

Today I managed to slim down ARM layers drastically (https://github.com/brefphp/aws-lambda-layers/pull/33). Cold starts are slightly improved for ARM runtimes.

Still it doesn't seem like PHP runs faster on ARM, but my benchmarks are very limited. Real apps would be much more useful here.

Also one last thing: APCu is now provided in ARM layers. That was the last missing thing.

---

Moving forward for Bref v2, do you think ARM layers should be marked as "beta", or as "stable"? Given the situation, I'd feel more comfortable with marking them as "beta", but if some of you start using them and are happy with them, we could mark them stable.

**Comment by SerkanYildiz** [comment]
Hi @mnapoli,
Thanks for your great work. I've an application running on aws lambda  / bref & symfony. The latter prevents me from installing the development version, therefore I couldn't test this. If you've plans to support (or I can do something to help you here) symfony with ARM layers, lmk. Thanks all

**Comment by mnapoli** [comment]
@SerkanYildiz If you have requirement conflicts, a quick and dirty option is:

```json
        "bref/bref": "v2.x-dev as 1.99.99"
```

**Comment by mnapoli** [comment]
I pushed some documentation in https://github.com/brefphp/bref/commit/ceb4e640b83682b65efdb1ce33983865fb782658.

I also tagged Bref 2.0.0-beta1, which lets you use the ARM layers: https://github.com/brefphp/bref/releases/tag/2.0.0-beta1

I will also tag new versions of the Symfony and Laravel bridge so that you can install the beta without conflicts.

I'll be closing this one, please open new issues if you see any issues with them! Thanks all for your involvement!

**Comment by mnapoli** [comment]
And an update just 2 hours later: it turns out we'll be able to support PHP 8.1 with Graviton too 🎉

Coming up in https://github.com/brefphp/aws-lambda-layers/pull/40

**Comment by shouze** [comment]
Support of php 8.2 on Graviton still miss me but part of the game, about 6 months to wait for now, not so long.

**Comment by shouze** [comment]
@aran112000 very good initiative, I’ve upvoted.

**Comment by pesseyjulien** [comment]
Hi, 

thanks for the work !

Quick question, do we also need to add 'arm' to the extra layer. 

Exemple : 

`${bref-extra:redis-php-81}` to `${bref-extra:redis-arm-php-81}` ?

Thanks in advance for your help,
Julien

**Comment by mnapoli** [comment]
@pesseyjulien ARM extensions aren't built yet (it will take a few weeks probably, we need to solve a few things first), and yes I do think we need to rebuild all of them for ARM unfortunately.

**Comment by eexit** [comment]
Hello,

Would it be possible to provide the ARM version of the Docker `bref/fpm-dev-gateway` image?

Thanks!

**Comment by mnapoli** [comment]
@eexit This image will no longer be useful in v2. I am working on the v2 blog post, but in the meantime here is a link that should explain what to do: https://github.com/brefphp/bref/blob/v2/docs/upgrading/v2.md#development-docker-images

> Also, why changing the image name for different arch? DockerHub natively handles multi-arch images...

Interesting question 🤔

Originally the images/layers were to be _very_ different, so having different Docker images made sense. Now I get why you are asking.

I could imagine the scenario where someone deploys to x86, and wants to run that env locally on Mac M1 (for example) -> they don't want to run the ARM image locally.

I'm not 100% sure that what I'm saying is a good argument, we could still consider merging all images into one, feel free to weigh in everyone. I don't think it will block v2, we can do that later.

**Comment by eexit** [comment]
Hello @mnapoli,

Thanks for your quick answer! I spent last evening trying to improve the performance of my PHP stack on my Mac M1, so that's the reason I'm asking 😃 

> Originally the images/layers were to be very different, so having different Docker images made sense. Now I get why you are asking.
> 
> I could imagine the scenario where someone deploys to x86, and wants to run that env locally on Mac M1 (for example) -> they don't want to run the ARM image locally.

Docker automatically pulls the right arch:

> When running an image with multi-platform support, docker automatically selects the image that matches your OS and architecture.

https://docs.docker.com/build/building/multi-platform/

This means, you can have the very same image name, this prevents per-arch stack config 💌 .

I'll give a shot to the `:2` image then.
Thanks!

**Comment by mnapoli** [comment]
>Docker automatically pulls the right arch:

Yes but that's what I wanted to avoid, because you'd be working locally with the ARM image and deploying to x86 Lambda.

**Comment by shouze** [comment]
> > Docker automatically pulls the right arch:
> 
> Yes but that's what I wanted to avoid, because you'd be working locally with the ARM image and deploying to x86 Lambda.

@mnapoli I'm not agree. It's nice & docker-ish way to have multi arch support for a given image, and under the same tag. Tag names shouldn't be used to specify an arch.
In the same time, it's the team decision to choose if depending of the local development/developer laptop arch, you want to force to use a sticky arch or let docker dynamically use x86 or arm64 depending of the arch of this developer laptop. And it works by design into docker.

For example, in `docker-compose.yaml`:

1. If you want to pin the platform:
```yaml
version: '3.9'
   services:
      foo:
         platform: linux/amd64
         image: bref/php:82
         
```
1. If you want to let docker dynamically use the platform of developer laptop (if arch is available in docker multi arch image):
```yaml
version: '3.9'
   services:
      foo:
         platform: linux/amd64
         image: bref/php:82
         
```

**Comment by georgeboot** [comment]
Agreed. They images probably should be very similar, so it shouldn't matter what platform you're on locally. If you care, set the `--platform linux/amd64` flag on the cli or the corresponding flag in the docker compose file.

**Comment by mnapoli** [comment]
OK, that's interesting, thanks all for chiming in.

So if I sum it up, it would make sense to drop all the `bref/arm-xxx` images and instead publish cross-platform images?

We don't foresee any downside (except having to set `--platform linux/amd64` explicitly for those who want to force a platform locally)?

**Comment by jasonmccallister** [comment]
@mnapoli that makes sense, but one thing that might be strange is that ARM support for multiple PHP versions is limited (unless that changed) so someone trying to use ARM and deploying `bref/php:82` might be surprised when there is no tag for the ARM platform?

**Comment by mnapoli** [comment]
FYI "multi-platform images" won't be implemented in the 2.0 release today (see https://github.com/brefphp/aws-lambda-layers/issues/62#issuecomment-1467790855 for more details, it's a bit harder than I thought), but I will work on this afterward.

The AWS Lambda layers will be perfectly fine, and running images locally will work, it will just be a bit slower until we have the multi-platform optimization.
